decentraland/unity-explorer
# Feat: Livekit retries and analytics

## What does this PR change?

HealthChecks
- HealthChecks are designed to ensure that the game could connect to the server, one of the use cases is to allow or don't allow a player to pass the loading screen
- Checks both statuses of urls Gatekeeper and Archipelago, then it tries to connect to both livekit room via IRoomHub
- Each failure attempt is tracked by Analytics
- Retries 3 times before failure
- Supports emulation of bad connection for debugging purposes
- By design IHealthCheck could be easily reused and extended for other use cases

Analytics
- Added LIVEKIT_HEALTH_CHECK_FAILED event, it requires integration testing
- On each failure attempt to connect to the server it sends an event
- AnalyticsExtensions, it allows to append an analytic decorator with nice to read Builder pattern style 

Splash Screen
- Fixed SplashAnimatorController.controller, previously it had 2 triggers and used only 1 of them, triggers are replaced by "Enable" boolean
- Introduced ISplashScreen to simplify splash screen usage, previously it passed in many placed by 2 objects: Gameobject and Animator, now it's a single object with an easy interface: ShowSplash() and HideSplash()

IUserInAppInitializationFlow
- since more and more operations are added to the flow there is introduced IStartupOperation interface, it allows to reduce complexity of RealUserInAppInitializationFlow and encapsulate only specific objects to the specific operations
- IStartupOperation provides a result from its execution and allows to customize flow with a certain case of failure
- In future it requires a popup that would show the error message from IStartupOperation to describe the reason of failure for an user

Debug settings
- Introduced IDebugSettings that exposes readonly properties to read
- Method Get() deleted due it could be easily forgotten and instead a member of team could pick a public field of IDebugSettings
- All fields are encapsulated as private, use properties instead


Multiplayer
- Multiplayer asmdef is subdivided to smaller ones to avoid cyclic dependencies because it had too many dependencies at time that it easily could invoke such case
- IConnectiveRoom's method Start() is replaced by StartAsync() to support awaiting a connection result to the server
- IRoomHub provides bool result that displays  success or failure of connection of both rooms

DecentralandUrls 
- Added 2 new urls: ArchipelagoStatus and GatekeeperStatus

Miscs
- Added nullables support for AsyncLoadReporting, Analytics asmdef
- Added NewPlayAudioScope to UIAudioEventsBus, it allows to play Audio only during the specific executing scope via IDisposable
- class Avatar, IsSameAvatar method is encapsulated to the object 
- added StatusCodeAsync to WebRequest adapter


## How to test the changes?

1. Launch the explorer
2. Tweak debug option EnableEmulateNoLivekitConnection and see the different results on SplashScreen
3. Ensure that the analytics event sends fine

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
